File: apps/api/app/schemas/product.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field


class RegisterRequest(BaseModel):
    email: EmailStr
    password: str = Field(min_length=8, max_length=128)
    full_name: str = Field(min_length=1, max_length=200)


class LoginRequest(BaseModel):
    email: EmailStr
    password: str


class UserOut(BaseModel):
    id: UUID
    email: EmailStr
    full_name: str
    is_active: bool


class CreatorOut(BaseModel):
    id: UUID
    display_name: str
    creator_type: str | None
    niche: str | None
    bio: str | None
    primary_language: str | None
    timezone: str | None
    location: str | None
    onboarding_step: int
    onboarding_completed: bool
    onboarding_data: dict[str, Any]


class CreatorUpdate(BaseModel):
    display_name: str | None = None
    bio: str | None = None
    niche: str | None = None
    primary_language: str | None = None
    timezone: str | None = None
    location: str | None = None


class OnboardingPayload(BaseModel):
    step: int = Field(ge=1, le=6)
    data: dict[str, Any]
    complete: bool = False


class AudienceBlock(BaseModel):
    description: str
    pain_points: list[str]
    interests: list[str]


class ContentPillarBlock(BaseModel):
    name: str
    description: str
    percentage: int = Field(ge=0, le=100)


class CreatorProfileSchema(BaseModel):
    niche: str
    positioning: str
    audience: AudienceBlock
    brand_personality: list[str]
    tone: str
    content_pillars: list[ContentPillarBlock]
    content_strategy: str
    recommended_formats: list[str]
    recommended_platforms: list[str]
    initial_strategy: list[str]


class GeneratedProfileOut(BaseModel):
    status: str
    profile: CreatorProfileSchema
    model: str | None = None
    prompt_version: str | None = None


class ProfileUpdateRequest(BaseModel):
    profile: CreatorProfileSchema
    accept: bool = False


class ContentGenerateRequest(BaseModel):
    topic: str | None = None
    platform: str | None = None
    format: str | None = None
    goal: str | None = None


class ContentAssetOut(BaseModel):
    id: UUID
    title: str
    script: str | None
    caption: str | None
    hashtags: list[Any]
    status: str
    scheduled_at: datetime | None
    asset_type: str


class ScheduleRequest(BaseModel):
    scheduled_at: datetime
    platform: str = "manual"


class CommentCreate(BaseModel):
    author_name: str
    body: str
    platform: str | None = None


class CommentReply(BaseModel):
    reply: str


class SettingsUpdate(BaseModel):
    approval_mode: str = Field(pattern="^(MANUAL|REVIEW_REQUIRED|AUTO_APPROVE)$")


class AssistantAsk(BaseModel):
    message: str = Field(min_length=1, max_length=4000)


class YouTubeConnectRequest(BaseModel):
    url: str = Field(min_length=2, max_length=500)

File: apps/web/lib/api/product.ts
import { apiGet, apiPost, apiPut } from "./client";

export type User = {
  id: string;
  email: string;
  full_name: string;
  is_active: boolean;
};

export const authApi = {
  me: () => apiGet<User>("/api/v1/auth/me"),
  register: (body: { email: string; password: string; full_name: string }) =>
    apiPost<User>("/api/v1/auth/register", body),
  login: (body: { email: string; password: string }) => apiPost<User>("/api/v1/auth/login", body),
  logout: () => apiPost<{ status: string }>("/api/v1/auth/logout"),
};

export type Creator = {
  id: string;
  display_name: string;
  creator_type: string | null;
  niche: string | null;
  bio: string | null;
  onboarding_step: number;
  onboarding_completed: boolean;
  onboarding_data: Record<string, unknown>;
};

export type YouTubeVideo = {
  video_id: string;
  title: string;
  url: string;
  views: number | null;
  views_label: string | null;
  published_label: string | null;
  thumbnail_url: string | null;
};

export type YouTubeSnapshot = {
  title: string;
  handle: string | null;
  url: string;
  description: string;
  subscriber_count: number | null;
  subscriber_label: string | null;
  video_count: number | null;
  videos: YouTubeVideo[];
  progress: { current: number; next_milestone: number; percent: number };
  next_actions: { id: string; title: string; detail: string }[];
  note: string;
  synced_at: string | null;
};

export type YouTubeConnection = {
  connected: boolean;
  platform: string;
  snapshot: YouTubeSnapshot | null;
};

export type Profile = {
  niche: string;
  positioning: string;
  audience: { description: string; pain_points: string[]; interests: string[] };
  brand_personality: string[];
  tone: string;
  content_pillars: { name: string; description: string; percentage: number }[];
  content_strategy: string;
  recommended_formats: string[];
  recommended_platforms: string[];
  initial_strategy: string[];
};

export const productApi = {
  creator: () => apiGet<Creator>("/api/v1/creators/me"),
  onboarding: () =>
    apiGet<{ step: number; completed: boolean; data: Record<string, unknown> }>("/api/v1/onboarding"),
  saveOnboarding: (body: { step: number; data: Record<string, unknown>; complete?: boolean }) =>
    apiPost("/api/v1/onboarding", body),
  generateProfile: () => apiPost<{ status: string; profile: Profile }>("/api/v1/onboarding/generate-profile"),
  strategy: () => apiGet<{ status: string; profile: Profile | null }>("/api/v1/strategy"),
  youtube: () => apiGet<YouTubeConnection>("/api/v1/platforms/youtube"),
  connectYouTube: (url: string) => apiPost<YouTubeConnection>("/api/v1/platforms/youtube", { url }),
  refreshYouTube: () => apiPost<YouTubeConnection>("/api/v1/platforms/youtube/refresh"),
  saveProfile: (profile: Profile, accept = false) =>
    apiPut<{ status: string; profile: Profile }>("/api/v1/strategy", { profile, accept }),
  dashboard: () => apiGet<Record<string, unknown>>("/api/v1/dashboard"),
  generateContent: (body: Record<string, string | undefined>) =>
    apiPost("/api/v1/content/generate", body),
  assets: () => apiGet<Array<Record<string, unknown>>>("/api/v1/content/assets"),
  ideas: () => apiGet<Array<Record<string, unknown>>>("/api/v1/content/ideas"),
  schedule: (id: string, scheduled_at: string, platform: string) =>
    apiPost(`/api/v1/content/assets/${id}/schedule`, { scheduled_at, platform }),
  approve: (id: string) => apiPost(`/api/v1/content/assets/${id}/approve`),
  analytics: () => apiGet<Record<string, unknown>>("/api/v1/analytics/overview"),
  comments: () => apiGet<Array<Record<string, unknown>>>("/api/v1/community/comments"),
  addComment: (body: { author_name: string; body: string; platform?: string }) =>
    apiPost("/api/v1/community/comments", body),
  replyComment: (id: string, reply: string) =>
    apiPost(`/api/v1/community/comments/${id}/reply`, { reply }),
  brand: () => apiGet<Record<string, unknown>>("/api/v1/brand-center"),
  settings: () => apiGet<{ approval_mode: string }>("/api/v1/settings"),
  saveSettings: (approval_mode: string) => apiPut("/api/v1/settings", { approval_mode }),
  assistantHistory: () => apiGet<Array<{ role: string; content: string }>>("/api/v1/assistant/messages"),
  askAssistant: (message: string) => apiPost<{ reply: string }>("/api/v1/assistant/ask", { message }),
};
